File: ios/Classes/utils/OperationManager.swift
class OperationManager {
    static let shared = OperationManager()
    
    internal var operations: [String: DispatchWorkItem] = [:]
    internal let queue = DispatchQueue(label: "com.easyvideoeditor.operation", attributes: .concurrent)
    
    private init() {}

    /// Generate a unique `operationId`  
    func generateOperationId() -> String {
        return UUID().uuidString
    }

    /// Register a new operation with its ID and work item
    func registerOperation(id: String, workItem: DispatchWorkItem) {
        queue.async(flags: .barrier) {
            self.operations[id] = workItem
        }
    }

    /// Cancel a specific operation by its ID
    func cancelOperation(_ id: String) {
        queue.async(flags: .barrier) {
            if let workItem = self.operations[id] {
                workItem.cancel()
                self.operations.removeValue(forKey: id)
            }
        }
    }

    /// Cancel all operations
    func cancelAllOperations() {
        queue.async(flags: .barrier) {
            for (_, workItem) in self.operations {
                workItem.cancel()
            }
            self.operations.removeAll()
        }
    }
}

File: android/src/main/kotlin/com/example/easy_video_editor/utils/OperationManager.kt
package com.example.easy_video_editor.utils

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.isActive
import kotlinx.coroutines.cancel
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * Singleton class that manages active operations and allows for cancellation
 */
object OperationManager {
    private val activeScopes = ConcurrentHashMap<String, CoroutineScope>()
    
    /**
     * Generates a unique operation ID
     */
    fun generateOperationId(): String {
        return UUID.randomUUID().toString()
    }
    
    /**
     * Registers an operation with its associated coroutine scope
     * 
     * @param id Unique identifier for the operation
     * @param scope Coroutine scope associated with the operation
     */
    fun registerOperation(id: String, scope: CoroutineScope) {
        activeScopes[id] = scope
    }
    
    /**
     * Cancels a specific operation
     * 
     * @param id Identifier of the operation to cancel
     * @return true if the operation was found and canceled, false otherwise
     */
    fun cancelOperation(id: String): Boolean {
        val scope = activeScopes[id]
        if (scope != null && scope.isActive) {
            scope.cancel("Operation canceled by user")
            activeScopes.remove(id)
            return true
        }
        return false
    }
    
    /**
     * Cancels all active operations
     * 
     * @return true if any operations were canceled, false if no operations were active
     */
    fun cancelAllOperations(): Boolean {
        if (activeScopes.isEmpty()) {
            return false
        }
        
        var anyCanceled = false
        activeScopes.forEach { (id, scope) ->
            if (scope.isActive) {
                scope.cancel("All operations canceled by user")
                anyCanceled = true
            }
            activeScopes.remove(id)
        }
        
        return anyCanceled
    }
}
